scandium-core/src/main/java/org/eclipse/californium/scandium/CookieGenerator.java

```diff
@@ -76,20 +107,39 @@ private Mac getHMAC() throws NoSuchAlgorithmException, CloneNotSupportedExceptio
 				hmac.init(generateSecretKey());
 			}
 			return (Mac) hmac.clone();
+		} catch (CloneNotSupportedException ex) {
+			// catch, if MacSpi is not cloneable (android)!
+			Mac hmac = this.hmac;
+			this.hmac = null;
+			return hmac;
 		} finally {
 			lock.writeLock().unlock();
 		}
 	}
 
+	/**
+	 * check, if {@link #lastSecretKey} has expired.
+	 * 
+	 * MUST be called in read or write scope of {@link #lock}!
+	 * 
+	 * @return {@code true}, if key is expired, {@code false} otherwise.
+	 */
 	private boolean isKeyExpired() {
-		return System.currentTimeMillis() - lastGenerationDate > keyLifetime;
+		return System.nanoTime() - lastGenerationNanos > keyLifetime;
```

Comment:
while we are at counting nanoseconds ...
we should not store the time we created the key but should calculate the time of expiration once only when we create the key and then simply compare `System.nanoTime()` to the expiration time. That way we can save the recurring calculation of the expiration time ...

Reply:
Just to say: my intention was to fix the issue in android. I had considered to just revert the previous optimization, but I assumed that this approach would cause less discussion then such a revert.
On that I was really wrong :-).

Reply:
I use nanoTime instead of currentTimeMillis, because nanoTime is not affected by system time adjustments. So dependencies on date time currentTimeMillis is the proper, dependencies on time periodes, nanoTime is proper. The resulution is not the concern.